File: web/src/gf_trader/ts/gf_trader.ts
/*
GloFlowTrader asset trading, management, and research platform
Copyright (C) 2019 Ivan Trajkovic

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

///<reference path="./../../d/jquery.d.ts" />

import "./gf_trader_watchlist";
import "./gf_trader_gemini";

namespace gf_trader {

$(document).ready(()=>{
    //-------------------------------------------------
    function log_fun(p_g,p_m) {
        var msg_str = p_g+':'+p_m
        //chrome.extension.getBackgroundPage().console.log(msg_str);

        switch (p_g) {
            case "INFO":
                console.log("%cINFO"+":"+"%c"+p_m,"color:green; background-color:#ACCFAC;","background-color:#ACCFAC;");
                break;
            case "FUN_ENTER":
                console.log("%cFUN_ENTER"+":"+"%c"+p_m,"color:yellow; background-color:lightgray","background-color:lightgray");
                break;
        }
    }
    //-------------------------------------------------
    gf_trader.init(log_fun);
});
//---------------------------------------------------
export function init(p_log_fun) {

    var stocks__open_bool = false;
    var crypto__open_bool = false;

    var stocks_container;
    var crypto_container;
    //----------------
    //STOCKS_WATCHLIST
    
    $('#markets #market__stocks_btn').on('click', ()=>{

        if (stocks__open_bool == false) {
            
            if (crypto__open_bool) {
                $(crypto_container).remove();
                crypto__open_bool = false;
            }

            stocks_container = gf_trader_watchlist.view(p_log_fun)
            $('body').append(stocks_container);
            stocks__open_bool = true;
        }
    });
    //----------------

    //----------------
    //CRYPTO_CURRENCY
    
    $('#markets #market__crypto_btn').on('click', ()=>{

        $(stocks_container).remove();
        if (crypto__open_bool == false) {

            if (stocks__open_bool) {
                $(stocks_container).remove();
                stocks__open_bool = false;
            }

            crypto_container = gf_trader_gemini.init(p_log_fun);
            
            crypto__open_bool = true;
        }
    });
    //----------------

    gf_trader_transactions.init__import(p_log_fun);
}
//---------------------------------------------------
export function http__get_symbols_daily_historic(p_symbols_lst :string[], p_onComplete_fun, p_onError_fun, p_log_fun) {

    const url_str = '/trader/quotes/daily_historic?symbols='+p_symbols_lst.join();
    p_log_fun('INFO','url_str - '+url_str);

    //-------------------------
    //HTTP AJAX
    $.get(url_str,
        function(p_data) {
            console.log('response received');
            //console.log('p_data - '+p_data);
            const data_map = JSON.parse(p_data);

            console.log('data_map["status_str"] - '+data_map["status_str"]);
            
            if (data_map["status_str"] == 'OK') {
                const quotes_lst = data_map['quotes_lst'];
                p_onComplete_fun(quotes_lst);
            }
            else {
                p_onError_fun(data_map["data"]);
            }
        });
    //------------------------- 
}
//---------------------------------------------------
export function http__get_symbols_latest(p_symbols_lst :string[], p_onComplete_fun, p_onError_fun, p_log_fun) {

    const url_str = '/trader/quotes/latest?symbols='+p_symbols_lst.join();
    p_log_fun('INFO','url_str - '+url_str);

    //-------------------------
    //HTTP AJAX
    $.get(url_str,
        function(p_data) {
            console.log('response received');
            //console.log('p_data - '+p_data);
            const data_map = JSON.parse(p_data);

            console.log('data_map["status_str"] - '+data_map["status_str"]);
            
            if (data_map["status_str"] == 'OK') {
                const quotes_lst = data_map['quotes_lst'];
                p_onComplete_fun(quotes_lst);
            }
            else {
                p_onError_fun(data_map["data"]);
            }
        });
    //------------------------- 
}
//---------------------------------------------------
}

File: go/gf_trader/gf_trader.go
/*
GloFlowTrader asset trading, management, and research platform
Copyright (C) 2019 Ivan Trajkovic

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

package main

//-------------------------------------------------
import (
	"fmt"
	"time"
	"net/http"
	"github.com/gloflow/gloflow/go/gf_core"
)

//-------------------------------------------------
type Runtime struct {
	Events_ctx  *gf_core.Events_ctx
	Runtime_sys *gf_core.Runtime_sys
}

//-------------------------------------------------
func main() {
	log_fun := gf_core.Init_log_fun()

	port_str := "4400"
	//-----------------
	//MONGODB

	mongodb_host_str    := "localhost"
	mongodb_db_name_str := "gf_trader"
	mongodb_db          := gf_core.Mongo__connect(mongodb_host_str, mongodb_db_name_str, log_fun)
	mongodb_coll        := mongodb_db.C("data")
	//-----------------

	runtime_sys := &gf_core.Runtime_sys{
		Service_name_str: "gf_trader",
		Log_fun:          log_fun,
		Mongodb_coll:     mongodb_coll,
	}
	
	events_ctx := gf_core.Events__init("/trader/events", runtime_sys)
	runtime    := &Runtime{
		Events_ctx:  events_ctx,
		Runtime_sys: runtime_sys,
	}
	
	//STATIC CONFIG 
	//symbols_lst := get_symbols()

	init_handlers(runtime)

	/*//------------
	//RUN_QUOTES_QUERY
	//this is for debugging, to view quotes HTTP query working and its output in the console
	quotes__get(symbols_lst,
			runtime,
			log_fun)
	//------------*/

	/*quotes__get_history(symbols_lst[0],
					log_fun)*/

	/*//START QUERIES_HTTP_UPDATE LOOP
	repeated__get_quotes(runtime,
					log_fun)*/

	account__create_defaults(runtime)

	//------------------------
	//DASHBOARD SERVING
	static_files__url_base_str := "/trader"
	gf_core.HTTP__init_static_serving(static_files__url_base_str, runtime_sys)
	//------------------------
	//CRYPTO_EXCHANGES
	gemini__init(runtime)
	binance__init(runtime)
	//------------------------
	//test()

	log_fun("INFO", ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>")
	log_fun("INFO", "STARTING HTTP SERVER - PORT - "+port_str)
	log_fun("INFO", "http://localhost:4400/trader/static/templates/gf_trader/gf_trader.html")
	log_fun("INFO", ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>")
	
	http_err := http.ListenAndServe(":"+port_str, nil)
	if http_err != nil {
		msg_str := "cant start listening on port - "+port_str
		log_fun("ERROR", msg_str)
		log_fun("ERROR", fmt.Sprint(http_err))
		panic(fmt.Sprint(http_err))
	}
}

//-------------------------------------------------
func market__is_open(p_runtime *Runtime) bool {
	p_runtime.Runtime_sys.Log_fun("FUN_ENTER", "gf_trader.market__is_open()")

	t := time.Now()

	//market not open over the weekend
	if t.Weekday().String() == "Saturday" || t.Weekday().String() == "Sunday" {
		return false
	} else {

		//market hasnt open yet, or has closed, for the day
		if (t.Hour() <= 9 && t.Minute() < 30.0) || t.Hour() >= 16 {
			return false
		}
	}

	return true
}